[ORKTextField] Coherent placeholder handling

rsanchezsaez opened the pull request:
This PR makes the handing of the placeholder in `ORKTextField` more similar to what `UITextField` does.

It comprises of two changes:
- Remove extra spacing between unit number and placeholder text. This makes the space between the unit and the placeholder to be consistent between *editing* and *non editing* modes.
- Do not hide placeholder when the editing text is empty (same as with `UITextField`).

Also reuses form item cells in `ORKStepViewController` so cells are not created twice when calculating their height.

---

Before: 

![simulator screen shot 9 sep 2015 02 46 13](https://cloud.githubusercontent.com/assets/444313/9751588/97db37b0-569e-11e5-9ff4-1c1c4db84219.png)
![simulator screen shot 9 sep 2015 02 59 19](https://cloud.githubusercontent.com/assets/444313/9751614/c7e3d2d2-569e-11e5-8ecd-3bfd16f7d54e.png)
![simulator screen shot 9 sep 2015 02 46 20](https://cloud.githubusercontent.com/assets/444313/9751587/97bcb8ee-569e-11e5-887b-ec4c7becb9a9.png)

---

After:

![simulator screen shot 9 sep 2015 03 05 42](https://cloud.githubusercontent.com/assets/444313/9751692/c7afd170-569f-11e5-9adf-66e63c4a933f.png)
![simulator screen shot 9 sep 2015 03 05 47](https://cloud.githubusercontent.com/assets/444313/9751695/cac923de-569f-11e5-9745-e6a45e0c5edc.png)
![simulator screen shot 9 sep 2015 03 05 53](https://cloud.githubusercontent.com/assets/444313/9751697/d037f0ac-569f-11e5-8d30-b0936dd7be66.png)

umerkhan-apple commented:
Did we want to extend this same functionality to the multiline textfields? For example, check out the *Message* form item under *Mini Form* in *ORKTest*.

rsanchezsaez commented:
@umerkhan-apple: I'll look into this in the following days.

umerkhan-apple commented:
@rsanchezsaez So did you get a chance to look at multiline textfields? Do we want to extend this functionality to them, or should we just leave it to single line textfields?

rsanchezsaez commented:
Hey Umer, 

Sorry, I missed the GitHub notification earlier.

I agree that the *placeholder* on *text views* (*multiline text fields*) should behave like the *text field* one. I'm looking into this.

rsanchezsaez commented:
@umerkhan-apple: Done, I have implemented the same placeholder behavior for `ORKAnswerTextView`.

umerkhan-apple commented:
@rsanchezsaez Awesome, I will test the changes shortly.

umerkhan-apple commented on ResearchKit/Common/ORKAnswerTextView.m:
```diff
@@ -30,44 +31,83 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
                                              selector:@selector(updateAppearance)
                                                  name:UIContentSizeCategoryDidChangeNotification
                                                object:nil];
+    [[NSNotificationCenter defaultCenter]  addObserver:self selector:@selector(textViewTextDidChange:)
```
Move selector argument onto a new line.

rsanchezsaez commented:
Adding the placeholder to the `ORKAnswerTextView` was tricky.

First, I tried using a plain `UILabel` but getting the positioning to be coherent with the typed in text was not easy.

So I settled to adding a non-interactive `UITextView` as a child of `ORKAnswerTextView`. For this to layout correctly on form steps, I needed to both set the frame manually (on layout subviews) and add constraints. I find this very weird because normally you either set the frame or use constraints, but never both. Removing one or the other makes it to stop being laid out correctly. If you have a suggestion for cleanest implementation, I'm all ears.

umerkhan-apple commented on ResearchKit/Common/ORKAnswerTextView.m:
```diff
@@ -30,44 +31,83 @@ OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 
+- (void)layoutSubviews {
+    [super layoutSubviews];
+    // Setting the frame directly causes a layout error on a form step (it looks like an iOS bug, as setting the frame should produce the same effect as settings the bounds and the center)
```
Typo: settings -> setting

umerkhan-apple commented:
@rsanchezsaez I agree that this is a bit of a tricky implementation. I will look into finding a cleaner approach if possible.

Currently, the framework is crashing if I change orientation from Portrait to Landscape inside `ORKTest` running `Mini Form` task.

##### Console log
> 2015-10-20 11:16:55.377 ORKTest[614:106204] -[NSIndexPath setExpectedLayoutWidth:]: unrecognized selector sent to instance 0x175dfda0 

_Tested on iPhone 5 and iPhone 6S._

rsanchezsaez commented:
I've fixed the crash, that was an unrelated silly mistake on my end.

Let me know if you come up with an alternate solution for the *text view placeholder*.

YuanZhu-apple commented:
<img width="315" alt="screen shot 2015-10-21 at 2 54 21 pm" src="https://cloud.githubusercontent.com/assets/11466704/10651299/acb0ef32-7803-11e5-94b6-1602288f4505.png">

Not form your code, I just noticed the textview is misaligned in formStep. Can you fix it in this PR, thanks!

YuanZhu-apple commented on ResearchKit/Common/ORKTextFieldView.m:
```diff
@@ -185,7 +184,7 @@ - (void)ork_setSuffix:(NSString *)suffix withColor:(UIColor *)color {
         return;
     }
-    _suffixLabel = [self ork_createTextLabelWithTextColor:color ?: [UIColor grayColor]];
+    _suffixLabel = [self ork_createTextLabelWithTextColor:(color ?: [UIColor ork_midGrayTintColor])];
```
Intention of this change?

rsanchezsaez replied:
This color is sometimes used for the *unit suffix label* when editing an `ORKTextFieldView` with no placeholder. You can see this if you revert this change, tap on the first text field of the first step on *Mini Form*, tap on the second field, and then tap on the first field again (the unit suffix label turns darker).

`ork_midGrayTintColor` is used as the color in all other `placeholder`s and *unit suffix* instances, so I figured it should be used here too.

But maybe we should also review the actual color-chosing logic, because it's weird that this color only gets used the second time you tap on the field but not the first one.

rsanchezsaez commented:
I'll later look into the alignment of the *text view*.

umerkhan-apple commented:
@rsanchezsaez After internal discussion, we have come to the following conclusion:
- When there is a unit specified on the text field, keep the behavior as it is in current RK.
- When there is no unit specified (basically, when it is functioning like a basic uitextfield) then show the placeholder until editing begins to stay consistent with iOS's uitextfield (your PR's implementation).

rsanchezsaez commented:
Ok. But why do that for first the case? Don't you think keeping the placeholder until some text is typed is more coherent and consistent?

umerkhan-apple commented:
The main reason for that conclusion was that having two placeholders seemed untidy and slightly confusing for the user.

rsanchezsaez commented:
But, in my opinion, the user perception is that, when editing an empty unit-suffixed *text field*, the unit label becomes part of the placeholder text and then, when they start typing, the "long placeholder" transforms to show only the unit suffix. I think this worked well after reducing the placeholder <> unit separation (see the before/after screenshots above). Don't you think?

If you remain unconvinced let me know and I'll update it to the behavior you requested. :)

YuanZhu-apple commented:
@rsanchezsaez I think the original set a more clear distinction between `placeholder` and `unit string`, see the margin between them when the textfield is inactive.

And when the textfield become active, showing both of them makes the field a little noisy.

[Original] 
<img width="223" alt="screen shot 2015-10-21 at 3 30 19 pm" src="https://cloud.githubusercontent.com/assets/11466704/10771822/26b8eae0-7cb0-11e5-86e3-d84d44167fb9.png">
<img width="267" alt="screen shot 2015-10-21 at 3 24 48 pm" src="https://cloud.githubusercontent.com/assets/11466704/10771821/26b4f08e-7cb0-11e5-97ae-aa37eaaef2bd.png">


[This PR] 
<img width="232" alt="screen shot 2015-10-21 at 3 32 20 pm" src="https://cloud.githubusercontent.com/assets/11466704/10771823/26bb7814-7cb0-11e5-9b04-b247d0c5b59b.png">
<img width="199" alt="screen shot 2015-10-21 at 3 15 17 pm" src="https://cloud.githubusercontent.com/assets/11466704/10771820/26a0fed0-7cb0-11e5-8d7d-c0678dfdd534.png">

rsanchezsaez commented:
I have updated this PR against the latest master, with the requested behavior:

- When there is a *suffix unit* specified on the text field, keep the behavior as it was.
- When there is no *suffix unit* specified, show the placeholder until the user writes text, to stay consistent with *iOS*'s `UITextField`.

I also added fixed a regression in which the user was being prevented leaving a form text field blank (a validation alert was being shown, you can try in `ORKTest`'s *Mini Form* by editing the first field and trying to move to the second without typing anything).

Let me know if you want additional changes. I think the animation glitch and the suffix color mismatch fix are worthwhile improvements as well.

rsanchezsaez commented:
Also, the *TextView* misalign shown in the above comment screenshot no longer seems to be an issue. Maybe *iOS 9.2* fixed it? I don't see any RK code change related to that.

YuanZhu-apple commented:
R=ME:) @umerkhan-apple Can you take a look as well?

YuanZhu-apple commented:
@umerkhan-apple Get a chance to try this?

rsanchezsaez commented:
@umerkhan-apple Ping? ;-)
